model-optimizer/extensions/ops/tensor_iterator.py

```diff
@@ -347,6 +347,51 @@ def ti_type_infer(node):
 
         ti_graph.remove_nodes_from([node.id for node in fake_input_const_nodes])
 
+    @staticmethod
+    def find_iterations_count_for_output(ti_node, output_rec):
+        def check_field(record, field):
+            return field in record and record[field] is not None
+        iterations_count = 1
+        # 1. check if we need concatenate iteration results for given output
+        if not check_field(output_rec, 'axis'):
+            # in this case we dont concat outputs, so iterations count is not needed really
+            return iterations_count
+
+        # 2. check if given output record contains values for 'end', so iterations count can be calculated from this record
+        if check_field(output_rec, 'end') and output_rec['end'] != -1 and \
+                check_field(output_rec, 'start') and output_rec['start'] != -1:
+            # get iterations count from output record
+            iterations_count = (output_rec['end'] - output_rec['start']) / output_rec['stride']
+            return iterations_count
+
+        # 3. find out iterations count from inputs.
+        # If no input contains 'axis' attribute then no slicing is in TI and it has only one iteration
+        # If several inputs have axis attribute with different iterations count then we use maximum value.
+        for in_rec in ti_node.input_port_map:
+            if not check_field(in_rec, 'axis'):
+                continue
+            if check_field(in_rec, 'end') and in_rec['end'] != -1 and \
+                    check_field(in_rec, 'start') and in_rec['start'] != -1:
+                in_rec_end = in_rec['end']
+                in_rec_start = in_rec['start']
+            elif check_field(in_rec, 'end') and in_rec['end'] != -1:
+                in_rec_end = in_rec['end']
+                in_rec_start = ti_node.in_port(in_rec['external_port_id']).data.get_shape()[in_rec['axis']]
+            elif check_field(in_rec, 'start') and in_rec['start'] != -1:
+                in_rec_end = ti_node.in_port(in_rec['external_port_id']).data.get_shape()[in_rec['axis']]
+                in_rec_start = in_rec['start']
+            else:
+                in_rec_end = ti_node.in_port(in_rec['external_port_id']).data.get_shape()[in_rec['axis']]
+                in_rec_start = 0
+            # in case of dynamic itreations count don't continue any calculations on this iteration
+            if not is_fully_defined(in_rec_end) or not is_fully_defined(in_rec_start):
+                iterations_count = dynamic_dimension_value
+                continue
```

Comment:
this logic has some mistakes:
1. we need to check `in_rec['stride']` existence in a record and only after that make computation of iteration_count
2. `in_rec['stride']` can be negative that means backward iterating and `iteration_count` will be different
3. we need to check existence of `start`, `end` in the map. and if some doesn't exist we need to deduce it using output shape only if `axis` exists

Reply:
The quite same logic should be implemented for output_port_map. We can re-use such code.